OrleansDashboard/History/TraceHistory.cs

```diff
@@ -0,0 +1,140 @@
+                    Grain = entry.Grain,
+                    Method = entry.Method,
+                    Period = now,
+                    SiloAddress = siloAddress
+                };
+
+                allGrainTrace.Add(grainTraceEntry);
+            }
+
+            // fill in any previously captured methods which aren't in this reporting window
+            var values = history.Where(x => string.Equals(x.SiloAddress, siloAddress, StringComparison.OrdinalIgnoreCase)).GroupBy(x => (x.Grain, x.Method)).Select(x => x.First());
+            foreach (var value in values)
+            {
+                if (!grainTrace.Any(x => 
+                    string.Equals(x.Grain, value.Grain, StringComparison.OrdinalIgnoreCase) &&
+                    string.Equals(x.Method, value.Method, StringComparison.OrdinalIgnoreCase)))
+                {
+                    allGrainTrace.Add(new GrainTraceEntry
+                    {
+                        Count = 0,
+                        ElapsedTime = 0,
+                        Grain = value.Grain,
+                        Method = value.Method,
+                        Period = now,
+                        SiloAddress = siloAddress
+                    });
+                }
+            }
+
+            history.AddRange(allGrainTrace);
+        }
+
+        public Dictionary<string, Dictionary<string, GrainTraceEntry>> QueryGrain(string grain)
+        {
+            var results = new Dictionary<string, Dictionary<string, GrainTraceEntry>>();
+
+            foreach (var historicValue in history.Where(x => x.Grain == grain))
+            {
+                var grainMethodKey = $"{grain}.{historicValue.Method}";
+
+                if (!results.TryGetValue(grainMethodKey, out var grainResults))
+                {
+                    results[grainMethodKey] = grainResults = new Dictionary<string, GrainTraceEntry>();
+                }
+
+                var key = historicValue.Period.ToPeriodString();
+
+                if (!grainResults.TryGetValue(grainMethodKey, out var value))
+                {
+                    grainResults[key] = value = new GrainTraceEntry
+                    {
+                        Grain = historicValue.Grain,
+                        Method = historicValue.Method,
+                        Period = historicValue.Period
+                    };
+                }
+
+                value.Count += historicValue.Count;
+                value.ElapsedTime += historicValue.ElapsedTime;
+                value.ExceptionCount += historicValue.ExceptionCount;
+            }
+            
+            return results;
+
+        }
+
+        public IEnumerable<TraceAggregate> GroupByGrainAndSilo()
+        {
+            return this.history.GroupBy(x => (x.Grain, x.SiloAddress)).Select(x => {
+                return new TraceAggregate{
+                    SiloAddress = x.Key.SiloAddress,
+                    Grain = x.Key.Grain,
+                    Count = x.Sum(y => y.Count),
```

Comment:
I know it is a micro-optimization, but a manual loop would be a little bit faster.

Reply:
good point, switched to a loop and benchmarking.

The 'higher order' programming part of my brain is arguing about it with the performance part :¬)